matflow/param_classes/seeds.py

```diff
@@ -1,53 +1,98 @@
+"""
+Schemes for describing seeds for growing crystals.
+"""
+from __future__ import annotations
 from dataclasses import dataclass
 from pathlib import Path
-from typing import Optional
+from typing import Any
+from typing_extensions import Self
 
 import numpy as np
-
+from numpy.typing import NDArray, ArrayLike
 from hpcflow.sdk.core.parameters import ParameterValue
 from matflow.param_classes.orientations import Orientations
 
 
 @dataclass
 class MicrostructureSeeds(ParameterValue):
+    """
+    The seeds for crystalline microstructure.
+    """
     _typ = "microstructure_seeds"
 
-    position: np.ndarray
-    box_size: np.ndarray
+    #: The positions of the seeds.
+    position: NDArray
+    #: The size of box containing the seeds.
+    box_size: NDArray
+    #: Label for the phase.
     phase_label: str
-    orientations: Optional[Orientations] = None
-    random_seed: Optional[int] = None
+    #: Orientation data.
+    orientations: Orientations | None = None
+    #: Seed for the random number generator, if used.
+    random_seed: int | None = None
 
-    def __post_init__(self):
+    def __post_init__(self) -> None:
         self.box_size = np.asarray(self.box_size)
         self.position = np.asarray(self.position)
-        if self.orientations:
-            if not isinstance(self.orientations, Orientations):
```

Comment:
I think we'll need to keep this. I have added some tests to `develop`, one of which will fail without this instance check.

Reply:
Need to accept a dict too? Fair enough.

Reply:
Found a neater way to handle this; `TypeGuard`. The trick there is that doesn't back-spill the types, which in turn means that we don't try to do weird things like constructing the type that is the intersection of several types that are disjoint in practice. I wish I didn't know so much about this aspect of Python's type system...